modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +171,113 @@ private void ensureOpen() {
     }
 
     private void reopenStreamOrFail(IOException e) throws IOException {
-        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
-            logger.warn(() -> format("""
-                failed reading [%s/%s] at offset [%s]""", blobStore.bucket(), blobKey, start + currentOffset), e);
-            throw e;
+        maybeRecordAndLogForRetry("reading", e);
+        maybeAbort(currentStream);
+        IOUtils.closeWhileHandlingException(currentStream);
+
+        maybeDelay();
+        openStreamWithRetry();
+    }
+
+    private <T extends Exception> void maybeRecordAndLogForRetry(String action, T e) throws T {
+        if (shouldRetry() == false) {
+            final var finalException = addSuppressedExceptions(e);
+            logForRetry(Level.WARN, action, finalException);
+            throw finalException;
         }
 
-        final int maxAttempts = blobStore.getMaxRetries() + 1;
+        attempt += 1;
+        // Log at info level for the 1st retry and every ~5 minutes afterward
+        logForRetry((attempt == 2 || attempt % 30 == 0) ? Level.INFO : Level.DEBUG, action, e);
+        if (failures.size() < MAX_SUPPRESSED_EXCEPTIONS) {
+            failures.add(e);
+        }
+    }
+
+    private void logForRetry(Level level, String action, Exception e) {
+        if (attempt == 1) {
```

Comment:
Should this be `attempt == 2`? Otherwise I think we'll not get there?

Reply:
It is `attempt == 1` because this branch is used to log failures where no retry should be attempted, i.e. repository analysis. I see now it can be confusing. So I separate this failure logging into a separate method and get rid of the check for `attempt`.